src/pw/fpga/opencl_utils.c

```diff
@@ -0,0 +1,370 @@
+/*****************************************************************************
+ *  CP2K: A general program to perform molecular dynamics simulations        *
+ *  Copyright (C) 2000 - 2019  CP2K developers group                         *
+ *****************************************************************************/
+
+/******************************************************************************
+ *  Author: Arjun Ramaswami
+ *****************************************************************************/
+
+#if defined ( __PW_FPGA )
+
+// global dependencies
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#define _USE_MATH_DEFINES
+#include <string.h>
+#include <sys/time.h>
+#include <stdarg.h>
+#include <stdbool.h>
+#include <unistd.h> // access in fileExists()
+#include "ctype.h"
+
+// common dependencies
+#include "CL/opencl.h"
+
+// function prototype
+static void tolowercase(char *p, char *q);
+static bool getBinaryPath(char *path, int N[3]);
+static char* loadBinary(char *binary_path, size_t *bin_size);
+
+// --- CODE -------------------------------------------------------------------
+
+/******************************************************************************
+ * \brief   return the first platform id with the name passed as argument
+ * \param   platform_name : search string
+ * \retval  din : platform id
+ *****************************************************************************/
+cl_platform_id findPlatform(char *platform_name){
+  unsigned int i;
+  cl_uint status;
+
+  // Check if there are any platforms available
+  cl_uint num_platforms;
+  status = clGetPlatformIDs(0, NULL, &num_platforms);
+  if (status != CL_SUCCESS){
+    printf("Query for number of platforms failed\n");
+    exit(0);
+  }
+
+  // Get ids of platforms available
+  cl_platform_id *pids = (cl_platform_id*) malloc(sizeof(cl_platform_id) * num_platforms);
+  status = clGetPlatformIDs(num_platforms, pids, NULL);
+  if (status != CL_SUCCESS){
+    printf("Query for platform ids failed\n");
+    exit(0);
```

Comment:
maybe use `1` here instead

Reply:
I let this remain as it is. Seems to be the norm in OpenCL codes as well as this makes it more readable since CL_SUCCESS resolves to 0

Reply:
The point here is that since this is running in the same address space as CP2K an `exit(0)` will terminate CP2K with an exit code of `0` which means a "successfull termination of the program" (see for example http://www.tldp.org/LDP/abs/html/exit-status.html). Which is clearly not the case.
Should this happen in a script then the script can only determine whether the run was successful (in the sense that CP2K finished gracefully, there is still the possibility that the calculation didn't converge) by checking the log file.

Reply:
Thank you for the explanation ..